pkgs/os-specific/linux/systemd/default.nix

```diff
@@ -160,6 +160,63 @@ stdenv.mkDerivation {
       --replace \
       "find_program('objcopy'" \
       "find_program('${stdenv.cc.bintools.targetPrefix}objcopy'"
+  '' + (let
+
+    # To get a list of dynamically loaded libraries issue something like
+    # `grep -ri 'dlopen("lib' $src` and update the below list.
+    dlopenLibs = [
```

Comment:
Right now we have no good way to override these. I'd like to be able to declare a `systemd-next` package by overriding systemd and providing new source and a (new?) set of patches. That will most likely also mean that we will have to provide new dlopen overrides.

One decision could be to just require those that override the package to invoke sed themselfs to fixup the new dlopen's in the prePatch phase. Opionions?

Reply:
How about instead of having an explicit list; we loop over everything in `buildInputs` implicitly?

Reply:
Basically this makes use handle `dlopen` exactly as any other build/runtime dependency

Reply:
We also do not really need to explicitly handle the case of `pkg = null;`  

As when the dependency is not there; the `dlopen("blah.so")` call will not be substituted; and `blah.so` won't be found already as `blah.so` is a relative path and won't be found .

Reply:
The dynamic linker defaults to searching the RPATH, LD_LIBRARY_PATH, … when the path does not contain (or starts with?) a `/`.

I would like to stick with making all those that we didn't define invalid store paths that never exist. This avoids weird situations where it sometimes works because LD_LIBRARY_PATH was set to something else (for whatever reason).

The problem with $buildInputs is that some of them do not have a lib folder but only declare their libs for some file in `nix-support`. Then you'll have to distinguish between runtime and build time dependencies (think cross compilation) and you are basically replicating stdenv's code for `findInputs`. This was my first approach but would end up being a lot more bash code than I actually would like to have in this derivation.

Reply:
> I would like to stick with making all those that we didn't define invalid store paths that never exist. This avoids weird situations where it sometimes works because LD_LIBRARY_PATH was set to something else (for whatever reason).

Setting LD_LIBRARY_PATH breaks all kinds of stuff not sure we need hardening against it. Afaik DT_RPATH as higher priority over `LD_LIBRARY_PATH`. So a valid path in the the rpath of the binary should have the same effect and can be only overridden by `LD_PRELOAD`. However patching source code seems less error prone to me compared to wrapping binaries since we don't know what source code ends up in some binaries.

Reply:
> The problem with $buildInputs is that some of them do not have a lib folder but only declare their libs via some file in nix-support. Then you'll have to distinguish between runtime and build time dependencies (think cross compilation) and walk the tree recursively. You are basically replicating stdenv's code for findInputs. This was my first approach but would end up being a lot more bash code than I actually would like to have in this derivation.

Fair enough.  This really gives me the feeling that having this source-code transformation is something that's useful to add to `stdenv`. I think it makes other applications using `dlopen` more pure too.

But that's for another PR.  

Do we want to do anything about override-ability? Or do we  consider that out of scope for now?

Reply:
IMHO if someone wants to work on a newer systemd version, that requires more than a few patch statements, that you can easily fit into prePatch, they should just fork nixpkgs for now. There is no point in incentivizing running (even more) downstream variants of systemd on NixOS beyond the current amount of customization. It also isn't like systemd is useless if you disable these new modificaitons. You might just run into weird runtime errors but those are then on you and not on us.